content/cosign-verify-bundles.md

```diff
@@ -0,0 +1,106 @@
+
+Just like before, we start by getting the artifact we want to verify:
+
+```
+$ curl -L https://github.com/cli/cli/releases/download/v2.54.0/gh_2.54.0_linux_armv6.tar.gz > gh_2.54.0_linux_armv6.tar.gz
+```
+
+Then we need the bundle containing SLSA provenance. Usually with Artifact Attestation you would use `gh attestation verify` to easily retrieve and verify the attestation at the same time, but here we're doing this the hard way to show how it's compatible with cosign.
+
+This release only publishes SLSA provenance, (i.e. it doesn't also publish a signed SBOM), so we can just get all the attestations with:
+
+```
+$ gh attestation download --repo cli/cli gh_2.54.0_linux_armv6.tar.gz
+Wrote attestations to file sha256:cd53809273ad6011fdd98e0244c5c2276b15f3dd1294e4715627ebd4f9c6e0f1.jsonl.
+```
+
+The verification command for public repositories is very similar to npm provenance, except of course we need to update the repository (and workflow) we expect the artifact to come from:
+
+```
+$ cosign verify-blob-attestation --bundle sha256:cd53809273ad6011fdd98e0244c5c2276b15f3dd1294e4715627ebd4f9c6e0f1.jsonl --new-bundle-format --certificate-oidc-issuer="https://token.actions.githubusercontent.com" --certificate-identity-regexp="^https://github.com/cli/cli/.github/workflows/deployment.yml.?" gh_2.54.0_linux_armv6.tar.gz
+Verified OK
+```
+
+Of course, if you're using a private GitHub repository things are a bit different. Instead of using the public good Sigstore instance (which includes the public transparency log Rekor, which you might not want to use for your private repository) we use a private Sigstore instance internal to GitHub. This instance has its own signing material (often called the trusted root) and instead of using the public Rekor instance it provides an independently observed time with a signed timestamp.
+
+So first we get the trusted root information for GitHub's internal Sigstore instance (I also didn't get this right on the first try):
+
+```
+$ gh attestation trusted-root | jq '.|select(.certificateAuthorities[0].uri=="fulcio.githubapp.com")' > github-trusted-root.json
+```
+
+And then we need to modify our verification command. We'll supply the correct trusted root, and we'll also tell cosign that we're using a signed timestamp instead of Rekor's Signed Certificate Timestamp (SCT). The flag is called `--insecure-ignore-sct` because it would be insecure if we weren't independently verifying the timestamp some other way. Here's what that verification command would look like:
+
+```
+$ cosign verify-blob-attestation --trusted-root github-trusted-root.json --use-signed-timestamps --insecure-ignore-sct --bundle ...
```

Comment:
Do you also need `--insecure-ignore-tlog`, since there is no Rekor proof? You can also use `--private-infrastructure`, these two flags are equivalent.

Reply:
We don't in this case since we are providing `--use-signed-timestamps`. If we did *not* have `--use-signed-timestamps`, we *would* need `--ignoreTlog`.

This was an implementation choice on how to map cosign flags to sigstore-go verifier config: https://github.com/sigstore/cosign/blob/b23586d6390d6a48ba4789848fe6ad89710afb7f/cmd/cosign/cli/verify/verify_bundle.go#L141-L147

I'll be the first to admit that this is a bit messy, and am open to feedback here.

Reply:
Whoops sorry, forgot this was a different path!

For https://github.com/sigstore/cosign/blob/b23586d6390d6a48ba4789848fe6ad89710afb7f/cmd/cosign/cli/verify/verify_bundle.go#L133-L135, I think we should remove `len(trustedroot.RekorLogs())` and require that you set a flag to skip the log check, otherwise a malicious can just remove the log entries from the bundle to skip over the check.

Reply:
I think this is okay - note that we're checking to see if there's Rekor entries in the trusted root, *not* the bundle.

For better or worse, we put a lot of trust in the supplied trusted root today (which is why we need to work towards having cosign retrieve it via TUF in more cases!)

Reply:
Oh you’re totally right, ignore this! Thanks